### ISSUE TITLE
Pyomo allows variable values to be lists

### ISSUE THREAD

**Issue by carldlaird** [Pyomo/pyomo#2041]
## Summary

Pyomo doesn't seem to validate the type of incoming data on variables. In the example below, I set the value of a variable to a list... (and it ... worked?)...

### Steps to reproduce the issue

```
import pyomo.environ as pyo
m = pyo.ConcreteModel()
m.x = pyo.Var()
m.x.fix([1,2,3])
print(pyo.value(m.x))
```
[1, 2, 3]

### Information on your system

Pyomo version: Pyomo 6.0.1 (devel {main})
Python version: 3.8.5
Operating system: OS X
How Pyomo was installed (PyPI, conda, source): source
Solver (if applicable):

**Comment by DLWoodruff** [comment]
I'm not sure this is a bug. The code does specify any domain for x.

**Comment by jsiirola** [comment]
This is definitely a bug (variable domains default to `Reals` if they are not specified):
```python
>>> m = ConcreteModel()
>>> m.x = Var()
>>> m.x = [1,2,3]
Traceback (most recent call last):
  ...
ValueError: Numeric value `[1, 2, 3]` (<class 'list'>) is not in domain Reals for Var x
>>> m.y = Var([1,2])
>>> m.y[1] = [1,2,3]
Traceback (most recent call last):
  ...
ValueError: Numeric value `[1, 2, 3]` (<class 'list'>) is not in domain Reals for Var y[1]
>>> m.x.fix([1,2,3])
>>> m.y[1].fix([1,2,3])
>>> m.pprint()
1 Set Declarations
    y_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    2 : {1, 2}

2 Var Declarations
    x : Size=1, Index=None
        Key  : Lower : Value     : Upper : Fixed : Stale : Domain
        None :  None : [1, 2, 3] :  None :  True :  True :  Reals
    y : Size=2, Index=y_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          1 :  None : [1, 2, 3] :  None :  True :  True :  Reals
          2 :  None :      None :  None : False :  True :  Reals

3 Declarations: x y_index y
```

The real challenge with variable domain validation is how to handle discrete (Binary, Integer) floating point values "within tolerances".  We generally allow that, as solvers regularly return floating point values and simply rounding is frequently the wrong thing to do.

**Comment by ghackebeil** [comment]
I agree that starting to check for inclusion in domains will likely open up a can of worms related to tolerances that we really don't want to open. Might make the most sense to simply enforce that values are numeric, but then again, maybe it's okay to let users set the value to whatever they want until it's a problem (e.g., a solver interface needs to use the variable value).

**Comment by jsiirola** [comment]
This was resolved by PEP #2180 (and PRs #2196 and #2214).  The reported case now raises an exception:
```python
>>> import pyomo.environ as pyo
>>> m = pyo.ConcreteModel()
>>> m.x = pyo.Var()
>>> m.x.fix([1,2,3])
Traceback (most recent call last):
  File "/home/jdsiiro/Research/pyomo/pyomo/core/expr/numvalue.py", line 119, in value
    obj.is_expression_type()
AttributeError: 'list' object has no attribute 'is_expression_type'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jdsiiro/Research/pyomo/pyomo/core/expr/numvalue.py", line 126, in value
    check_if_numeric_type_and_cache(obj)
  File "/home/jdsiiro/Research/pyomo/pyomo/core/expr/numvalue.py", line 401, in check_if_numeric_type_and_cache
    if obj_class is (obj + 0).__class__:
TypeError: can only concatenate list (not "int") to list

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jdsiiro/Research/pyomo/pyomo/core/base/var.py", line 267, in fix
    self.set_value(value, skip_validation)
  File "/home/jdsiiro/Research/pyomo/pyomo/core/base/var.py", line 371, in set_value
    val = value(val)
  File "/home/jdsiiro/Research/pyomo/pyomo/core/expr/numvalue.py", line 131, in value
    (type(obj).__name__,))
TypeError: Cannot evaluate object with unknown type: list
```
